c/httpserver.c

```diff
@@ -5503,61 +5747,218 @@ static int httpHandleTCP(STCBase *base,
       break; /* end server socket processing */
 
     } else {
-      zowelog(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_DEBUG3, "handle peer socket read\n");
+      bool shouldBreak = false;
+      handlerStatus = handlePeerSocketRead(base,module,socket,extension,&shouldBreak);
+      if (shouldBreak){
+        break;
+      }
+    } /* end peer socket handling */
+    
+  } while(0);
 
-      SocketExtension *peerExtension = extension;
-      HttpConversation *conversation = (HttpConversation*)extension->protocolHandler;
+  return handlerStatus;
+}
 
-      if (NULL == conversation) {
-        zowelog(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_DEBUG, "*** peerExtension protocolHandler is NULL ***\n");
-        /* we can't do a full conversation cleanup, just close the socket and unregister the socketExtension, leaks be damned */
-        socketClose(peerExtension->socket, &returnCode,&reasonCode);
-        handlerStatus = 8;
-        break; /* end peer socket processing */
-      }
+typedef struct PipeSocketEntry_tag{
+  int uniqueID;
+  Socket *demuxSocket;
+  Socket *handlerSocket;  /* used by http/tcp code */
+  Socket *muxSocket;      
+} PipeSocketEntry;
+
+static PipeSocketEntry *getOrMakePipeSocketEntry(HttpServer *server, 
+						 STCModule *module, 
+						 Socket *demuxSocket, 
+						 int uniqueID,
+						 int *pipeError){
+  zowelog(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_DEBUG2, "getOrMake=%d\n",uniqueID);
+  PipeSocketEntry *entry = htIntGet(server->syntheticPipeSockets,uniqueID);
+  zowelog(NULL, LOG_COMP_HTTPSERVER, ZOWE_LOG_DEBUG2, "ht result = 0x%p\n",entry);
+
+  if (entry == NULL){
+    int fdsToHttp[2] ={ -1, -1};  /* readEnd, writeEnd */
+    int fdsFromHttp[2] ={ -1, -1};
+    if (pipe(fdsToHttp)){
+      *pipeError = errno;
+      return NULL;
+    }
+    if (pipe(fdsFromHttp)){
+      *pipeError = errno;
```

Comment:
Should `fdsToHttp` be closed?

Reply:
Maybe, but in single user mode it will end naturally when zss ends.   It's not a system-wide resource, it's  process-level resource that auto-cleans.